cloudinit/sources/DataSourceOpenStack.py

```diff
@@ -128,7 +128,8 @@ def _get_data(self):
 
         if self.perform_dhcp_setup:  # Setup networking in init-local stage.
             try:
-                with EphemeralDHCPv4(self.fallback_interface):
+                with EphemeralDHCPv4(self.fallback_interface,
```

Comment:
I don't expect this will actually buy us anything in practice, EphemeralDHCPv4 is only called in OpenStackLocal (during init-local) during _get_data and _get_data  doesn't get called per-boot for OpenStack it only gets recrawled if instance-id changes.

That said, if behavior did change for OpenStack and we expected _get_data to be called or triggered per-boot or due to a hotplug event. then we probably should have this check in place to avoid additional unneeded dhcp setup cost.

Reply:
> Also, during the first run of _get_data, self.metadata_address is not yet set (because it gets determined by wait_for_metadata_service which is actually called later).

should we use the module constants instead?

Reply:
We can't use constants;  the user may have provided a datasource config to point to a different URL;

We would need to modify EphemeralDHCPv4 to accept a list of urls; and then the same set of URLs we pass to the wait_for_metadata_service() could be specified.

In general as @blackboxsw mentions; this does not help on first boot, or even subsequent boots.  On first boot, we're running before networking is up; so the conn check is going to fail and we'll bring up EphmeralDHCP.  On second boot; we won't crawl the IMDS data a second time unless our instance-id changes;  if this is a new instance; we're still running _before_ networking is enabled so we're still going to fail to connect and attempt to bring up EphemeralDHCP.